lib/elixir/lib/module.ex

```diff
@@ -1174,6 +1301,42 @@ defmodule Module do
     :ok
   end
 
+  defp set_doc_false_if_unset(table, unread_line) do
+    case :ets.lookup(table, :doc) do
+      [{:doc, {_line, <<_::binary>>}, _accumulated?, _unread_line}] ->
+        :doc_exists
+      [{:doc, {_line, false}, _accumulated?, _unread_line}] ->
+        :doc_already_set_false
+      [] ->
+        :ets.insert(table, {:doc, {unread_line, false}, false, unread_line})
+    end
+  end
+
+  defp warn_if_impl_argument_is_not_a_behaviour(module, behaviour) when is_atom(behaviour) do
+    cond do
+     behaviour in get_behaviours(module) ->
+      :ok
+     true ->
+      IO.warn "module attribute @impl was set but this module does not " <>
+              "implement #{inspect(behaviour)}. Did you forget to specify" <>
+              " @behaviour #{inspect(behaviour)} ?"
+    end
+  end
+
+  defp get_behaviours(module) do
+    get_attribute(module, :behaviour)
+  end
+
+  # TODO: address "module attribute @impl_required was set but never used" warnings
+  defp set_impl_required(table, unread_line) do
+    case :ets.lookup(table, :impl_required) do
```

Comment:
For internal attributes, such as `:impl_required`, you can namespace it as`{:elixir, :impl_required}` to avoid conflicts with user attributes.